src/coreclr/vm/excep.cpp

```diff
@@ -2460,7 +2455,15 @@ void StackTraceInfo::SaveStackTrace(BOOL bAllowAllocMem, OBJECTHANDLE hThrowable
                     }
                 }
 
-                ((EXCEPTIONREF)ObjectFromHandle(hThrowable))->SetStackTrace(gc.stackTrace.Get(), gc.dynamicMethodsArray);
+                if (gc.dynamicMethodsArray != NULL)
+                {
+                    gc.dynamicMethodsArray->SetAt(0, gc.stackTrace.Get());
```

Comment:
The stacktrace is visible to other threads at this point. It is too late to set the dynamicMethodArray here.

Here is a repro for this race condition:
1. Add sleep in front of this line to make a deterministic simple repro possible:
```diff
                 if (gc.dynamicMethodsArray != NULL)
                 {
+                    ClrSleepEx(2000,FALSE);
+
                     gc.dynamicMethodsArray->SetAt(0, gc.stackTrace.Get());
                     ((EXCEPTIONREF)ObjectFromHandle(hThrowable))->SetStackTrace(dac_cast<OBJECTREF>(gc.dynamicMethodsArray));
                 }
```
2. Run https://gist.github.com/jkotas/12fdbd933a1ad0c1cc32a96353c4b1d4 (on checked build)
3. You should see it to fail with:
```
Assert failure(PID 20804 [0x00005144], Thread: 13800 [0x35e8]): handle

CORECLR! LCGMethodResolver::GetManagedResolver + 0x2C (0x00007ff8`4f3f1aac)
CORECLR! DebugStackTrace::GetStackFramesInternal + 0xF80 (0x00007ff8`4f3dd3f0)
```